File: loss/MultiClassCrossEntropy.m
classdef MultiClassCrossEntropy < BaseLoss
    % Multi-class cross entropy
    % References:
    % https://www.ics.uci.edu/~pjsadows/notes.pdf
    
    properties
    end
    
    methods (Access = 'public')
        function [loss, gradients] = GetLossAndGradients(obj, prob, targets)
            N = size(prob,1);
            % Considering that the scores are already converted to
            % probabilities.
            % Get the indexes of the correct classes
            [~, idxCorrect] = max(targets,[],2);
            
            % Get the probabilities of the correct classes
            probCorrect = diag(prob(:,idxCorrect));

            loss = -sum(log(probCorrect))/N;            
            
            % Get gradient of loss w.r.t to the correct score
            gradients = prob;
            gradients_correct = probCorrect - 1;
            % Put the calculated correction back on 
            gradients(sub2ind(size(gradients),[1:length(idxCorrect)]',idxCorrect)) = gradients_correct;            
            gradients = gradients / N;
        end
    end    
end

File: utils/optimization_functions/cppPrj/loss/multiclasscrossentropy.cpp
#include "multiclasscrossentropy.h"

MultiClassCrossEntropy::MultiClassCrossEntropy()
{

}

tuple<float, Tensor<float> > MultiClassCrossEntropy::GetLossAndGradients(const Tensor<float> &scores, const Tensor<float> &targets){
    auto N = scores.GetRows();
    //Considering that the socres are already converted to probabilities (After softmax activation)

    // Get the indexes of the correct classes
    auto argMaxTargets = MathHelper<float>::MaxTensor(targets,1);
    auto idxCorrect = argMaxTargets.second;

    // Get the probabilities of the correct class
    auto selScoresTargets = scores.EltWiseMult(targets);
    auto probCorrect = MathHelper<float>::GetNonZero(selScoresTargets);

    // Calculate the loss
    auto loss = -MathHelper<float>::SumVec(MathHelper<float>::Log(probCorrect))/N;

    // Get the gradient of the loss w.r.t
    auto gradients = scores;
    auto gradients_correct = probCorrect - 1;

    // Update on each row of gradients (using the index from idxCorrect) the correct gradient (gradients_correct)
    auto rowsGradients = gradients.GetRows();
    auto idxCorrectIt = idxCorrect.begin();
    auto gradients_correct_it = gradients_correct.begin();
    for (auto rows = 0; rows < rowsGradients; ++rows){
        // Select
        gradients(rows, *idxCorrectIt) = *gradients_correct_it;
        gradients_correct_it++;
        idxCorrectIt++;
    }
    // Take the effect of the batch size
    gradients = gradients / N;

    return make_tuple(loss,gradients);
}
